compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt

```diff
@@ -5,32 +5,131 @@
 import org.jetbrains.kotlin.fir.symbols.impl.FirPropertyAccessorSymbol
+import org.jetbrains.kotlin.fir.symbols.impl.FirPropertySymbol
 import org.jetbrains.kotlin.lexer.KtTokens
 
 // See old FE's [DeclarationsChecker]
 object FirMemberPropertyChecker : FirRegularClassChecker() {
     override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
-        for (member in declaration.declarations) {
-            if (member is FirProperty) {
-                checkProperty(declaration, member, context, reporter)
+        val memberPropertySymbols = declaration.declarations.filterIsInstance<FirProperty>().map { it.symbol }.toSet()
+
+        // A property is known to be initialized only if it is initialized
+        //   1) with its own initializing expression;
+        //   2) at every class constructor;
+        //   3) at any of class's anonymous initializers; or
+        //   4) at other property's initializing expression
+
+        // 2) Property can be initialized at constructors. Since it's unknown what constructor will be used, the property can be determined
+        // as initialized only if it is initialized at every constructor. We should consider a delegated constructor, e.g.,
+        //   constructor() { x = ... }
+        //   constructor(...): this() { ... }  // x will be initialized via this() delegation
+
+        // 3) Property can be initialized at any of class's anonymous initializers (all of initializers will be executed), e.g.,
+        //   init { x = ... }
+        //   ...
+        //   init { y = ... }
+
+        // 4) Property can be initialized at other property's initializing expression too, e.g.,
+        //   val initX = inlineMe { x = ... } // where inlineMe returns the value of the last expression of the lambda
+
+        // Here, we accumulate initialized info of a property per constructor, anonymous initializer, or property initialization
+        val initializedInConstructor =
+            mutableMapOf<FirPropertySymbol, EventOccurrencesRange>().withDefault { EventOccurrencesRange.ZERO }
+        val initializedInInitOrOtherProperty =
+            mutableMapOf<FirPropertySymbol, EventOccurrencesRange>().withDefault { EventOccurrencesRange.ZERO }
+
+        // To handle the delegated constructor call, we need a cache from constructor to (analyzed) property init info.
+        val constructorToData =
+            mutableMapOf<FirConstructor, PropertyInitializationInfo>().withDefault { PropertyInitializationInfo.EMPTY }
+
+        fun collectInfoFromGraph(
+            graph: ControlFlowGraph,
+            map: MutableMap<FirPropertySymbol, EventOccurrencesRange>,
+            acc: (EventOccurrencesRange, EventOccurrencesRange) -> EventOccurrencesRange,
+            delegatedConstructor: FirConstructor? = null,
+        ) {
+            val delegatedInfo = delegatedConstructor?.let { constructorToData.getValue(it) } ?: PropertyInitializationInfo.EMPTY
```

Comment:
You'll get an issue with constructors that are not topologically sorted:
```kt
class Z {
    val a: Int
    constructor(x: Int): this()
    constructor() {
        a = 1
    }
}
```

Info for delegatedConstructor (2nd constructor) will be empty when you collect info for the 1st constructor.

Reply:
Good point! I also updated a similar test to make sure we visit constructors in a topological order.